craft_parts/executor/part_handler.py

```diff
@@ -0,0 +1,384 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2017-2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Definitions and helpers for part handlers."""
+
+import logging
+import os
+import os.path
+import shutil
+from glob import iglob
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional
+
+from craft_parts import callbacks, errors, packages, plugins, sources
+from craft_parts.actions import Action, ActionType
+from craft_parts.infos import PartInfo, StepInfo
+from craft_parts.packages import errors as packages_errors
+from craft_parts.parts import Part
+from craft_parts.plugins import Plugin
+from craft_parts.state_manager import states
+from craft_parts.state_manager.states import StepState
+from craft_parts.steps import Step
+from craft_parts.utils import os_utils
+
+from .organize import organize_files
+from .step_handler import StepContents, StepHandler
+
+logger = logging.getLogger(__name__)
+
+
+class PartHandler:
+    """Handle steps for a part using the appropriate plugins.
+
+    :param part: The part being processed.
+    :param part_info: Information about the part being processed.
+    :param part_list: A list containing all parts.
+    """
+
+    def __init__(
+        self,
+        part: Part,
+        *,
+        part_info: PartInfo,
+        part_list: List[Part],
+    ):
+        self._part = part
+        self._part_info = part_info
+        self._part_list = part_list
+
+        self._plugin = plugins.get_plugin(
+            part=part,
+            properties=part.plugin_properties,
+            part_info=part_info,
+        )
+
+        self._part_properties = part.spec.marshal()
+        self._source_handler = sources.get_source_handler(
+            application_name=part_info.application_name,
+            part=part,
+            project_dirs=part_info.dirs,
+        )
+
+        self._build_packages = _get_build_packages(part=self._part, plugin=self._plugin)
```

Comment:
these could be lazy loaded
https://docs.python.org/3.8/library/functools.html#functools.cached_property

Is there a reason this is done during __init__? If so, maybe code comment it so no one tries to optimize. 
The other option deviating from lazy loading with the same objective is to just call `_get_build_packages(part=self._part, plugin=self._plugin)` in the assets dict decl.